### ISSUE TITLE
Decouple Google API client creation logic to support individual clients for different users

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#3658]
Currently Site Kit only uses one instance of a `Google_(Site_Kit_)Client`, since all requests are made on behalf of the current user. In the future, dashboard sharing will enable access delegation, in which for certain modules and read actions requests can be made on behalf of another user than the current one. When this happens, multiple different client instances may need to be used in combination.

The logic to currently create a new Site Kit client with all the configuration that is needed is currently baked into the `OAuth_Client` class, which creates and uses a single client instance throughout. We should keep this overall approach as is, having the `OAuth_Client` class always act on behalf of the current user - actions like granting scopes do not work without involving the user that owns the credentials anyway. However, we need to decouple the logic to generate a new client, so that we can also instantiate clients based on other existing read credentials.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new class `Google\Site_Kit\Core\Authentication\Clients\Client_Factory` should be introduced with a public static method `create_client` which creates a new configured `Google_Site_Kit_Client` based on given arguments.
    * Dependencies like the `Options` or `User_Options` instance should be passed to the method.
    * The client should be set up exactly like the client currently is in `OAuth_Client::setup_client`, except in a more dynamic way, i.e. relying on the above dependencies instead of class properties.
    * This should e.g. allow instantiating a client based on data for a different user than the current user.
* The `OAuth_Client` class should have all related logic that is now moved to the new class removed and instead use the new class itself.

## Implementation Brief

* 

### Test Coverage

* 

### Visual Regression Changes

* 

## QA Brief

* 

## Changelog entry

*

**Comment by aaemnnosttv** [comment]
Thanks @felixarntz – the IB looks great overall. A few thoughts I have to consider:

> Deprecate `get_access_token`, `set_access_token`, `get_refresh_token`, `set_refresh_token`, and make them all use `get_token` / `set_token` respectively.

No objections to this really, but looking at the changes in the PR, almost all of our usage of the access token are to check that it's not empty to see if the user is authenticated for some condition. I would suggest we add a dedicated method for checking this on the `OAuth_Client`, maybe `has_token() : bool`. Also `get_access_token` in particular might be worth keeping as not deprecated since we do need the actual token in a few places and it is a fair bit cleaner (and safer) than the conditional that's needed every time otherwise for safely accessing the value in the token array that is returned. Alternatively, we could consider always returning an array with all the keys in all cases, just with empty values, rather than just an empty array.

On a similar note, what do you think about making the token its own class? We have similar classes for most other settings and options so it's a bit odd that this one would be an exception where a dedicated class could help manage some of the complexity around it, especially since the value is sourced from multiple individual settings.

> Split the `OAuth_Client` class, moving all the pieces that are also crucial in case of a secondary (i.e. not the current user) into a new `OAuth_Inactive_Client` class which the `OAuth_Client` class then extends.

The `OAuth_Inactive_Client` name sounds a bit off to me. Maybe something like `OAuth_Client_Delegated` would be more clear, and might even need methods which are unique to this case as well, in which case maybe an abstract `Base_OAuth_Client` would make sense if needed?

---

Again, overall this looks good to continue with. Let me know what you think about the above 👍

**Comment by felixarntz** [comment]
@aaemnnosttv I like all those ideas. I've updated the IB accordingly:
* No longer deprecate `get_access_token` since it makes sense to keep - only the other ones.
* Introduce a `Token` class representing the four token values in one place, since they belong together.
* Introduce only an abstract `Base_OAuth_Client` for now, in preparation for the future, since we don't _yet_ need any delegated client class.

**Comment by aaemnnosttv** [comment]
@felixarntz Great! A few final thoughts:

> While these encompass multiple user options, the class can technically still extend `User_Setting`, which allows it to be used in the same way as if these four values were stored in the database together (which _might_ even make sense to change eventually, but not here).

I don't think extending `User_Setting` would be possible without changing the way it is stored since all of the methods rely on the single `OPTION` constant in the class. Also, `User_Setting` takes a single `User_Options_Interface` which requires different implementations for the internal values (access/refresh tokens are encrypted while expiration and created_at timestamps are not). I think for now we'd need to make `Token` its own class that takes `User_Options` and `Encrypted_User_Options` instances to provide to the 4 separate classes we'd need for the key-value pairs.

In this regard, one might think it would actually be simpler to change the way it's stored but I think the implications of that change would require too much to include here. I think that's worth doing in the future though.

With that said, I think implementing `Token` as a standalone class would be fine for now, with temporary new `User_Setting` classes for the 4 key-values due to the need for both `User_Options` implementations. We should probably also add tests for this which is not yet listed in the current IB.

Regarding the names of the `OAuth_Client` variants, I used both prefixed and suffixed names in my last comment, but I would prefer if we use suffixed names for class variants to keep them grouped together. E.g. `OAuth_Client_Base` and `OAuth_Client_Delegated` over the `*_OAuth_Client` alternatives. I know the idea is not to add the delegated variant yet but saying as a general convention to establish which is already somewhat present with the existing classes we have.

**Comment by felixarntz** [comment]
@aaemnnosttv Fair point about `Token` not extending `User_Setting`. I've adjusted the IB to just implement it as an independent class for now - still following the same method signatures though, in preparation for the future.

Regarding 4 `User_Setting` classes for the individual user options, I don't think that's useful at this point. They have no real value in addition to how it is now, particularly because they'll likely be temporary. We don't register those user settings so far, and starting to do that now would be confusing given we likely want to drop them. I've clarified that in the IB and scoped it to centrally handling the four user options in `Token`, which is closer to what we eventually want to have. The 4 classes would just make for a little extra work IMO not worth it.

About using suffixes for the class names, I agree - I've updated the IB with that as well.

**Comment by aaemnnosttv** [comment]
> Regarding 4 `User_Setting` classes for the individual user options, I don't think that's useful at this point. They have no real value in addition to how it is now, particularly because they'll likely be temporary. We don't register those user settings so far, and starting to do that now would be confusing given we likely want to drop them.

@felixarntz Sounds good! We can just use `User_Options` internally without any extra classes for the parts like you said.

One question I have left that would be good to clarify I think would be around the constructor/dependencies given to `Token`. Since it would need both `User_Options` and `Encrypted_User_Options` (prefix 😄 ) would you want to provide both as constructor dependencies or only the first and instantiate the encrypted instance internally? Also, do you think `Token` should implement `User_Aware_Interface` or leave that for the dependencies only?

**Comment by felixarntz** [comment]
@aaemnnosttv I think it makes sense to only pass `User_Options` into the constructor and instantiate `Encrypted_User_Options` based on it internally. I've added a point to the IB.

Regarding `User_Aware_Interface`, let's not implement it in `Token` for now, since I think we won't need it. If we need it later we can always add it, or once it eventually becomes a `User_Setting` it would by definition have it anyway. I'm not strictly opposed to adding it, but it's also not necessary, so let's leave it out.

**Comment by aaemnnosttv** [comment]
Great, thanks!

IB ✅

**Comment by wpdarren** [comment]
@felixarntz @aaemnnosttv I would like to make sure I am understanding this section in the QAB:

`Ensure that refreshing an access token still works (for this you'll need to ensure the current access token is considered expired, which you can easily accomplish by manually updating the database user option googlesitekit_access_token_expires_in to something like "1", leading to it expiring 1 second after the original creation time).`

So, I changed the token expiry to 1 second but what should I see on my test site?  I am still connected to Site Kit with my Google account and all of the modules are still connected.  Is that expected? Sorry if its a silly question 😃

**Comment by aaemnnosttv** [comment]
@wpdarren sure I can explain – when you go through the oAuth flow (signing in with Google) and land back in Site Kit, you get an access token which is what we need to make requests with to the Google APIs. This is only valid for 1 hour though, after which it's automatically refreshed on-demand later when another request is made. So what that part of the QAB is describing is a way to manually expire the access token so that it would be refreshed on the next request. The next request is key there – if all you do is set the expiration and refresh, it's likely that no request would be made due to cache in the browser. So after setting the expiration, clear your session storage and then refresh again to ensure requests will be made, and then if everything works then the token should have been refreshed. It would be good to check the token expiration was updated though which should be back to (or close to) `3600`. I can provide commands for you to do that if you'd like, or you could always pair with an engineer to cover this part specifically 👍

**Comment by wpdarren** [comment]
@aaemnnosttv brilliant explanation, thank you!  

I  _might_ be good with the commands but will reach out to an engineer if I'm struggling.

**Comment by wpdarren** [comment]
### QA Update: Pass ✅ 

- Can confirm that going through the plugin setup works as expected. 
- When resetting `wp_googlesitekit_access_token_expires_in` to `1` and clearing cache then refreshing, the token was reset to `3599`
